// mouse-firework/dist/index.esm.js
import anime from 'theme-shokax-anime';

const sample = (raw) => {
    return Array.isArray(raw) ? anime.random(raw[0], raw[1]) : raw;
};
const hasAncestor = (node, name) => {
    name = name.toUpperCase();
    do {
        if (node === null || node === undefined)
            break;
        if (node.nodeName === name)
            return true;
    } while ((node = node.parentNode) !== null);
    return false;
};
const setEndPos = (p, particle) => {
    var _a;
    if (particle.move.includes("emit")) {
        let { emitRadius = [50, 180] } = (_a = particle.moveOptions) !== null && _a !== void 0 ? _a : {};
        const angle = (anime.random(0, 360) * Math.PI) / 180;
        emitRadius = sample(emitRadius);
        const radius = [-1, 1][anime.random(0, 1)] * emitRadius;
        p.endPos = {
            x: p.x + radius * Math.cos(angle),
            y: p.y + radius * Math.sin(angle),
        };
    }
};
const setEndRotation = (p, particle) => {
    var _a;
    if (particle.move.includes("rotate")) {
        const { angle = [-180, 180] } = (_a = particle.moveOptions) !== null && _a !== void 0 ? _a : {};
        p.endRotation = sample(angle);
    }
};

class BaseEntity {
    constructor(ctx, x, y, color, radius, alpha, lineWidth) {
        this.ctx = ctx;
        this.x = x;
        this.y = y;
        this.color = color;
        this.radius = radius;
        this.alpha = alpha;
        this.lineWidth = lineWidth;
        this.rotation = 0;
    }
    draw() {
        const { ctx, x, y } = this;
        ctx.save();
        ctx.translate(x, y);
        ctx.rotate(this.rotation * (Math.PI / 180));
        ctx.globalAlpha = this.alpha;
        this.paint();
        if (this.lineWidth) {
            ctx.lineWidth = this.lineWidth;
            ctx.strokeStyle = this.color;
            ctx.stroke();
        }
        else {
            ctx.fillStyle = this.color;
            ctx.fill();
        }
        ctx.globalAlpha = 1;
        ctx.restore();
    }
}

class Circle extends BaseEntity {
    paint() {
        this.ctx.beginPath();
        this.ctx.arc(0, 0, this.radius, 0, 2 * Math.PI);
    }
}

class Polygon extends BaseEntity {
    constructor(ctx, x, y, color, radius, alpha, sides, lineWidth) {
        super(ctx, x, y, color, radius, alpha, lineWidth);
        this.sides = sides;
    }
    paint() {
        const { ctx, sides, radius } = this;
        ctx.beginPath();
        ctx.moveTo(radius * Math.cos(0), radius * Math.sin(0));
        for (let i = 1; i <= sides; i++) {
            const angle = (i * 2 * Math.PI) / sides;
            ctx.lineTo(radius * Math.cos(angle), radius * Math.sin(angle));
        }
        ctx.closePath();
    }
}

class Star extends BaseEntity {
    constructor(ctx, x, y, color, radius, alpha, spikes, lineWidth) {
        super(ctx, x, y, color, radius, alpha, lineWidth);
        this.spikes = spikes;
    }
    paint() {
        const { ctx, spikes, radius } = this;
        ctx.beginPath();
        ctx.moveTo(0, 0 - radius);
        for (let i = 0; i < spikes * 2; i++) {
            const angle = (i * Math.PI) / spikes - Math.PI / 2;
            const length = i % 2 === 0 ? radius : radius * 0.5;
            const px = Math.cos(angle) * length;
            const py = Math.sin(angle) * length;
            ctx.lineTo(px, py);
        }
        ctx.closePath();
    }
}

const createCircle = (ctx, x, y, particle) => {
    const num = sample(particle.number);
    let { radius, alpha = 100, lineWidth, } = particle.shapeOptions;
    if (Array.isArray(alpha)) {
        alpha = alpha.map((a) => a * 100);
    }
    else {
        alpha *= 100;
    }
    const circles = [];
    for (let i = 0; i < num; i++) {
        const p = new Circle(ctx, x, y, particle.colors[anime.random(0, particle.colors.length - 1)], sample(radius), sample(alpha) / 100, sample(lineWidth));
        setEndPos(p, particle);
        setEndRotation(p, particle);
        circles.push(p);
    }
    return circles;
};
const createStar = (ctx, x, y, particle) => {
    const num = sample(particle.number);
    let { radius, alpha = 100, lineWidth } = particle.shapeOptions;
    if (Array.isArray(alpha)) {
        alpha = alpha.map((a) => a * 100);
    }
    else {
        alpha *= 100;
    }
    const spikes = sample(particle.shapeOptions.spikes);
    const stars = [];
    for (let i = 0; i < num; i++) {
        const p = new Star(ctx, x, y, particle.colors[anime.random(0, particle.colors.length - 1)], sample(radius), sample(alpha) / 100, spikes, sample(lineWidth));
        setEndPos(p, particle);
        setEndRotation(p, particle);
        stars.push(p);
    }
    return stars;
};
const createPolygon = (ctx, x, y, particle) => {
    const num = sample(particle.number);
    let { radius, alpha = 100, lineWidth, } = particle.shapeOptions;
    if (Array.isArray(alpha)) {
        alpha = alpha.map((a) => a * 100);
    }
    else {
        alpha *= 100;
    }
    const polygons = [];
    const sides = sample(particle.shapeOptions.sides);
    for (let i = 0; i < num; i++) {
        const p = new Polygon(ctx, x, y, particle.colors[anime.random(0, particle.colors.length - 1)], sample(radius), sample(alpha) / 100, sides, sample(lineWidth));
        setEndPos(p, particle);
        setEndRotation(p, particle);
        polygons.push(p);
    }
    return polygons;
};

const canvasEl = document.createElement("canvas");
canvasEl.style.cssText =
    "position:fixed;top:0;left:0;pointer-events:none;z-index:9999999";
document.body.appendChild(canvasEl);
const ctx = canvasEl.getContext("2d");
const tap = /iPhone|iPad|iPod|Android/i.test(navigator.userAgent)
    ? "touchstart"
    : "click";
let pointerX = 0;
let pointerY = 0;
const setCanvasSize = () => {
    canvasEl.width = document.documentElement.clientWidth * 2;
    canvasEl.height = document.documentElement.clientHeight * 2;
    canvasEl.style.width = document.documentElement.clientWidth + "px";
    canvasEl.style.height = document.documentElement.clientHeight + "px";
    const ctx = canvasEl.getContext("2d");
    ctx.setTransform(1, 0, 0, 1, 0, 0);
    ctx.scale(2, 2);
};
const updateCoords = (e) => {
    pointerX =
        e.clientX ||
            (e.touches && e.touches[0].clientX);
    pointerY =
        e.clientY ||
            (e.touches && e.touches[0].clientY);
};
const setParticleMovement = (particle) => {
    var _a, _b, _c, _d;
    const { move } = particle;
    let dist = {};
    if (move.includes("emit")) {
        const { radius = 0.1, alphaChange = false, alphaEasing = "linear", alphaDuration = [600, 800], } = (_a = particle.moveOptions) !== null && _a !== void 0 ? _a : {};
        let { alpha = 0 } = (_b = particle.moveOptions) !== null && _b !== void 0 ? _b : {};
        if (Array.isArray(alpha)) {
            alpha = alpha.map((a) => a * 100);
        }
        else {
            alpha *= 100;
        }
        let alphaOptions = {};
        if (alphaChange) {
            alphaOptions = {
                alpha: {
                    value: sample(alpha) / 100,
                    easing: alphaEasing,
                    duration: sample(alphaDuration),
                },
            };
        }
        dist = {
            ...dist,
            x: (p) => p.endPos.x,
            y: (p) => p.endPos.y,
            radius: sample(radius),
            ...alphaOptions,
        };
    }
    else if (move.includes("diffuse")) {
        const { diffuseRadius = [80, 160], lineWidth = 0, alphaEasing = "linear", alphaDuration = [600, 800], } = (_c = particle.moveOptions) !== null && _c !== void 0 ? _c : {};
        let { alpha = 0 } = (_d = particle.moveOptions) !== null && _d !== void 0 ? _d : {};
        if (Array.isArray(alpha)) {
            alpha = alpha.map((a) => a * 100);
        }
        else {
            alpha *= 100;
        }
        dist = {
            ...dist,
            radius: sample(diffuseRadius),
            lineWidth: sample(lineWidth),
            alpha: {
                value: sample(alpha) / 100,
                easing: alphaEasing,
                duration: sample(alphaDuration),
            },
        };
    }
    if (move.includes("rotate")) {
        dist = {
            ...dist,
            rotation: (p) => p.endRotation,
        };
    }
    return dist;
};
const renderParticle = (targets) => {
    for (const target of targets) {
        target.draw();
    }
};
const render = anime({
    duration: Infinity,
    update() {
        ctx.clearRect(0, 0, canvasEl.width, canvasEl.height);
    },
});
let currentCallback = null;
let globalOptions = null;
const initFireworks = (options) => {
    globalOptions = options;
    if (currentCallback) {
        document.removeEventListener(tap, currentCallback, false);
    }
    currentCallback = (e) => {
        for (const excludeElement of options.excludeElements) {
            if (hasAncestor(e.target, excludeElement)) {
                return;
            }
        }
        render.play();
        updateCoords(e);
        animateParticles(pointerX, pointerY);
    };
    document.addEventListener(tap, currentCallback, false);
    setCanvasSize();
    window.removeEventListener("resize", setCanvasSize, false);
    window.addEventListener("resize", setCanvasSize, false);
};
const animateParticles = (x, y) => {
    if (!globalOptions)
        return;
    const { particles } = globalOptions;
    const timeLine = anime().timeline();
    for (const particle of particles) {
        const { duration, easing } = particle;
        let targets = [];
        if (particle.shape === "circle") {
            targets = createCircle(ctx, x, y, particle);
        }
        else if (particle.shape === "star") {
            targets = createStar(ctx, x, y, particle);
        }
        else if (particle.shape === "polygon") {
            targets = createPolygon(ctx, x, y, particle);
        }
        const dist = setParticleMovement(particle);
        timeLine.add({
            targets,
            duration: sample(duration),
            easing,
            update: renderParticle,
            ...dist,
        });
    }
    timeLine.play();
};
var index = (options) => {
    if (document.readyState === "loading") {
        window.addEventListener("DOMContentLoaded", () => initFireworks(options), {
            passive: true,
        });
    }
    else {
        initFireworks(options);
    }
};

export { index as default };
