// unlazy/dist/unlazy.js
const y = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7", p = typeof window > "u", b = !p && "loading" in HTMLImageElement.prototype, w = !p && (!("onscroll" in window) || /(gle|ing|ro)bot|crawl|spider/i.test(navigator.userAgent));
function z(e, t = document) {
  return typeof e == "string" ? [...t.querySelectorAll(e)] : e instanceof Element ? [e] : [...e];
}
function v(e, t) {
  let s;
  return function(...c) {
    s && clearTimeout(s), s = setTimeout(() => {
      s = void 0, e(...c);
    }, t);
  };
}
function L(e = 'img[loading="lazy"]', {
  hash: t = !0,
  hashType: s = "blurhash",
  placeholderSize: c = 32,
  updateSizesOnResize: n = !1,
  onImageLoad: i
} = {}) {
  const a = /* @__PURE__ */ new Set();
  for (const r of z(e)) {
    const l = u(r, { updateOnResize: n });
    if (n && l && a.add(l), !r.dataset.src && !r.dataset.srcset) {
      (typeof __UNLAZY_LOGGING__ > "u" || __UNLAZY_LOGGING__) && console.error("[unlazy] Missing `data-src` or `data-srcset` attribute", r);
      continue;
    }
    if (w || !b) {
      S(r), A(r), d(r);
      continue;
    }
    if (r.src || (r.src = y), r.complete && r.naturalWidth > 0) {
      f(r, i);
      continue;
    }
    const E = () => f(r, i);
    r.addEventListener("load", E, { once: !0 }), a.add(
      () => r.removeEventListener("load", E)
    );
  }
  return () => {
    for (const r of a)
      r();
    a.clear();
  };
}
function O(e = 'img[data-sizes="auto"], source[data-sizes="auto"]') {
  for (const t of z(e))
    u(t);
}
function f(e, t) {
  const s = new Image(), { srcset: c, src: n, sizes: i } = e.dataset;
  if (i === "auto") {
    const a = _(e);
    a && (s.sizes = `${a}px`);
  } else
    e.sizes && (s.sizes = e.sizes);
  c && (s.srcset = c), n && (s.src = n), s.addEventListener("load", () => {
    S(e), A(e), d(e), t == null || t(e);
  });
}
const o = /* @__PURE__ */ new WeakMap();
function u(e, t) {
  var c;
  if (e.dataset.sizes !== "auto")
    return;
  const s = _(e);
  if (s && (e.sizes = `${s}px`), ((c = e.parentElement) == null ? void 0 : c.tagName.toLowerCase()) === "picture" && !(t != null && t.skipChildren) && [...e.parentElement.getElementsByTagName("source")].forEach(
    (n) => u(n, { skipChildren: !0 })
  ), t != null && t.updateOnResize) {
    if (!o.has(e)) {
      const n = v(() => u(e), 500), i = new ResizeObserver(n);
      o.set(e, i), i.observe(e);
    }
    return () => {
      const n = o.get(e);
      n && (n.disconnect(), o.delete(e));
    };
  }
}
function d(e) {
  e.dataset.src && (e.src = e.dataset.src, e.removeAttribute("data-src"));
}
function A(e) {
  e.dataset.srcset && (e.srcset = e.dataset.srcset, e.removeAttribute("data-srcset"));
}
function S(e) {
  const t = e.parentElement;
  (t == null ? void 0 : t.tagName.toLowerCase()) === "picture" && ([...t.querySelectorAll("source[data-srcset]")].forEach(A), [...t.querySelectorAll("source[data-src]")].forEach(d));
}
function _(e) {
  var t, s;
  return e instanceof HTMLSourceElement ? (s = (t = e.parentElement) == null ? void 0 : t.getElementsByTagName("img")[0]) == null ? void 0 : s.offsetWidth : e.offsetWidth;
}
const R = Object.freeze({
  autoSizes: O,
  lazyLoad: L,
  loadImage: f
});
var h;
(h = document.currentScript) != null && h.hasAttribute("init") && L();
export {
  O as autoSizes,
  R as default,
  L as lazyLoad,
  f as loadImage
};
